active/NumberBox/NumberBox.md

```diff
@@ -0,0 +1,220 @@
+only when there's a bug in the caller, such as argument exception.  But if for some
+reason it's necessary for a caller to catch an exception from an API, call that
+out with an explanation either here or in the Examples -->
+
+## API Notes
+<!-- Option 1: Give a one or two line description of each API (type
+and member), or at least the ones that aren't obvious
+from their name.  These descriptions are what show up
+in IntelliSense. For properties, specify the default value of the property if it
+isn't the type's default (for example an int-typed property that doesn't default to zero.) -->
+
+<!-- Option 2: Put these descriptions in the below API Details section,
+with a "///" comment above the member or type. -->
+
+### Behavioral Components
+
+| Property | Notes |
+|:---:|:---|
+| Value/Text Changed | * When Text is changed via codebehind or by user input ("Enter" is pressed or the NumberBox loses focus) the NumberBox will be prompted to update its Text and Value properties. A TextChanged event will be raised that will exposes the new Text and the old Value and allows the developer to intercept the and manipulate these properties (such as for manually handling validation error). Value will then be updated to the Text input converted to a Double. <br><br> * Conversely, code behind updates to Value will raise a ValueChanged event that will exposes the new Value and the old Text and allows the developer to intercept the and manipulate these properties (such as for manually handling validation error). Text will then be updated to the Value input converted to a String.    |
+| Validation | * Following the [precedent of the Windows ecosystem](https://github.com/microsoft/microsoft-ui-xaml/issues/483#issuecomment-498485363), if invalid input is entered and the developer does not intercept it  via the ValueChanged or TextChanged events (which exposes the changed property and the one to be updated) to create valid input, the invalid property will be reverted to the other's preserved valid value. <br><br? * Disabling validation override will surface a vaidation error indicator and message to the user. <br><br> * Should MinValue%StepSize != 0 or MaxValue%StepSize != 0, this will also trigger a validation error that the developer may intercept during the TextChanged event. If validation override is not disabled and the change is not intercepted via the respective event, the value will be returned to the last valid state. E.g, in this scenario, if StepSize=0.2, MaxValue=3.0, and the value 2.9 is incremented, it will be returned to 2.9. |
+| StepSize vs. MaxValue/MinValue |  |
+
+## API Details
+<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->
+
+```c++ 
+runtimeclass NumberBoxValueChangedEventArgs
+{
+    String Text;
+    Double Value
+};
+
+runtimeclass NumberBoxTextChangedEventArgs
+{
+    String Text;
+    Double Value
+};
+
+unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.TextBox
```

Comment:
I think this class should derive from [`RangeBase`](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.Primitives.RangeBase), not `TextBox`

Reply:
> I think this class should derive from [`RangeBase`](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.Primitives.RangeBase), not `TextBox`

The control has more in common with the TextBox control than it does a Slider, ScrollBar or ProgressBar control.  There is also a wish for it to automatically benefit from Prefix and Suffix properties which could be added to the TextBox.

Reply:
I disagree. The main programmatic access to the control is through value bindings and user being able to interact with the value through text is more of an implementation/templating detail of the control. You can put a `TextBox` in the template and not even have to programmatically expose much of its interface. The control is mostly used for entering values and we should make it easily swappable with a slider, make sure that accessibility features provide value information etc. 

You can do it either way of course, but `RangeBase` has basically exactly the properties we need, so if we derive from `TextBox` - we'll be reinventing the wheel.

Reply:
> I disagree. The main programmatic access to the control is through value bindings and user being able to interact with the value through text is more of an implementation/templating detail of the control. You can put a `TextBox` in the template and not even have to programmatically expose much of its interface. The control is mostly used for entering values and we should make it easily swappable with a slider, make sure that accessibility features provide value information etc.
> 
> You can do it either way of course, but `RangeBase` has basically exactly the properties we need, so if we derive from `TextBox` - we'll be reinventing the wheel.

So the Textbox becomes a TemplatePart.  How would that work when adding the Spin Buttons and integrating them into the TextBox itself?

Reply:
Why do they have to be integrated into the `TextBox`? They can be next to it. You can also retemplate the `TextBox` and will likely need to do it either way.

Reply:
Actually prefer the up/down buttons to be next to the control instead of being integrated into the TextBox.

Reply:
> Why do they have to be integrated into the `TextBox`? They can be next to it. You can also retemplate the `TextBox` and will likely need to do it either way.

Making the buttons and TextBox integrated is part of the proposal, because separate buttons and a TextBox could be done with binding.

![image](https://user-images.githubusercontent.com/7389110/59051299-6f8bbc00-8884-11e9-8c14-6aa63940cdb6.png))

Reply:
Thanks, @xyzzer, we examined deriving from RangeBase and TextBox. As noted on the proposal and in the review summaries here, we have decided against deriving from any class in preference of containing TextBox to avoid aggregating unusable properties on this control. Your suggestions enabled us to have comprehensive discussions on this topic, so I want to be sure to thank you and credit you for our progress. 😄